### PR TITLE
ci: run compatibility tests between latest client and HEAD gateway

### PR DESCRIPTION
**Pull request by thomaseizinger**
Currently, our CI tests that the latest released client and gateway are compatible with the current portal. To allow for smooth upgrades of deployed infrastructure, we also need to test that any changes we are making to the gateway are compatible with the latest release of the client. This allows customers to upgrade their gateways ahead of time before we publish updates of the clients.

This PR adds a matrix to the compatibility tests to ensure just that.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
This will need a change of the required CI jobs to merge.

**Review by conectado — APPROVED** [review]

**Comment by thomaseizinger** [comment]
@jamilbk Please update the required checks at your convenience and then queue this!

**Comment by jamilbk** [comment]
@thomaseizinger Unfortunately I need to change the test name, because they change each time meaning required checks will never pass:

Screenshot 2024-06-16 at 2 52 13 PM

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -144,18 +144,30 @@ jobs:
...
+      fail-fast: false
+      matrix:
+        client:
+          - image: "ghcr.io/firezone/client"
+            tag: "latest"
+        gateway:
+          - image: ${{ needs.build-artifacts.outputs.gateway_image }}
+            tag: ${{ github.sha }}
+          - image: "ghcr.io/firezone/gateway"
+            tag: "latest"
```
Comment (jamilbk) [inline .github/workflows/ci.yml]:
```suggestion
          - image: ${{ needs.build-artifacts.outputs.gateway_image }}
            tag: ${{ github.sha }}
            ci-name: sha
          - image: "ghcr.io/firezone/gateway"
            tag: "latest"
            ci-name: latest
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -144,18 +144,30 @@ jobs:
...
     # Don't run compatibility tests when called from hotfix.yml or publish.yml on `main` because
     # it'll be red if there was a breaking change we're tring to publish,
     # and the deploy_production workflow checks for main to be green.
     if: ${{ github.event_name == 'pull_request' || github.event_name == 'merge_group' }}
+    name: compatibility-tests-client(${{ matrix.client.tag }})-gateway(${{ matrix.gateway.tag }})
```
Comment (jamilbk) [inline .github/workflows/ci.yml]:
```suggestion
    name: compatibility-tests-client(${{ matrix.client.tag }})-gateway(${{ matrix.gateway.ci-name }})
```

**Comment by thomaseizinger** [comment]
> @thomaseizinger Unfortunately I need to change the test name, because they change each time meaning required checks will never pass:
> Screenshot 2024-06-16 at 2 52 13 PM

Lol, yeah that was a bit of a silly idea 😂

**Pull request merged**